File: framec_tests/golang/state_vars/state_vars.go
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

package state_vars

import "golang/framelang"


func NewStateVars() StateVars {
    m := &stateVarsStruct{}
    
    // Validate interfaces
    var _ StateVars = m
    var _ StateVars_actions = m
    
    // Create and intialize start state compartment.
    m._compartment_ = NewStateVarsCompartment(StateVarsState_Init)
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type StateVarsState uint

const (
    StateVarsState_Init StateVarsState = iota
    StateVarsState_A
    StateVarsState_B
)

type StateVars interface {
    X() 
    Y() 
    Z() 
}

type StateVars_actions interface {
}


type stateVarsStruct struct {
    _compartment_ *StateVarsCompartment
    _nextCompartment_ *StateVarsCompartment
}

//===================== Interface Block ===================//

func (m *stateVarsStruct) X()  {
    e := framelang.FrameEvent{Msg:"X"}
    m._mux_(&e)
}

func (m *stateVarsStruct) Y()  {
    e := framelang.FrameEvent{Msg:"Y"}
    m._mux_(&e)
}

func (m *stateVarsStruct) Z()  {
    e := framelang.FrameEvent{Msg:"Z"}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *stateVarsStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case StateVarsState_Init:
        m._StateVarsState_Init_(e)
    case StateVarsState_A:
        m._StateVarsState_A_(e)
    case StateVarsState_B:
        m._StateVarsState_B_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *stateVarsStruct) _StateVarsState_Init_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        compartment := NewStateVarsCompartment(StateVarsState_A)
        compartment.StateVars["x"] = 0
        
        m._transition_(compartment)
        
        return
    }
}

func (m *stateVarsStruct) _StateVarsState_A_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "X":
        m._compartment_.StateVars["x"] = m._compartment_.StateVars["x"].(int) + 1
        
        return
    case "Y":
        compartment := NewStateVarsCompartment(StateVarsState_B)
        compartment.StateVars["y"] = 10
        
        compartment.StateVars["z"] = 100
        
        m._transition_(compartment)
        
        return
    case "Z":
        compartment := NewStateVarsCompartment(StateVarsState_B)
        compartment.StateVars["y"] = 10
        
        compartment.StateVars["z"] = 100
        
        m._transition_(compartment)
        
        return
    }
}

func (m *stateVarsStruct) _StateVarsState_B_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "X":
        compartment := NewStateVarsCompartment(StateVarsState_A)
        compartment.StateVars["x"] = 0
        
        m._transition_(compartment)
        
        return
    case "Y":
        m._compartment_.StateVars["y"] = m._compartment_.StateVars["y"].(int) + 1
        
        return
    case "Z":
        m._compartment_.StateVars["z"] = m._compartment_.StateVars["z"].(int) + 1
        
        return
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *stateVarsStruct) _transition_(compartment *StateVarsCompartment) {
    m._nextCompartment_ = compartment
}

func (m *stateVarsStruct) _do_transition_(nextCompartment *StateVarsCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions


********************************************************/

//=============== Compartment ==============//

type StateVarsCompartment struct {
    State StateVarsState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewStateVarsCompartment(state StateVarsState) *StateVarsCompartment {
    c := &StateVarsCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}

File: framec_tests/cpp/state_vars/state_vars.cpp
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

#include <vector>
#include <any>
using namespace std;
#include "../FrameLang/FrameLang.h"

//=============== Compartment ==============//

class StateVarsCompartment
{
public:
    int state;
    
    StateVarsCompartment(int state)
    {
        this->state = state;
    }
    
    std::unordered_map<std::string, std::any> stateArgs;
    std::unordered_map<std::string, std::any> stateVars;
    std::unordered_map<std::string, std::any> enterArgs;
    std::unordered_map<std::string, std::any> exitArgs;
    FrameEvent *_forwardEvent = nullptr;
};

class StateVars
{
public:
    StateVarsCompartment *_compartment_;
    StateVarsCompartment *_nextCompartment_;
    
    
    
public:
    StateVars()
    {
        // Create and intialize start state compartment.
        _state_ = static_cast<int>(StateVarsState::INIT);
        
        _compartment_ = new StateVarsCompartment(this->_state_);
        _nextCompartment_ = nullptr;
        
        
        // Send system start event
        FrameEvent frame_event(">", std::unordered_map<std::string, std::any>());
        _mux_(&frame_event);
    
    }
    
    // states enum
private:
    enum class StateVarsState
    {
        INIT = 0,
        A = 1,
        B = 2
    };
    
    //====================== Multiplexer ====================//

private:
    void _mux_(FrameEvent *e)
    {
        if(this->_compartment_->state == static_cast<int>(StateVarsState::INIT))
        {
            this->_sInit_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(StateVarsState::A))
        {
            this->_sA_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(StateVarsState::B))
        {
            this->_sB_(e);
        }
        
        
        if(this->_nextCompartment_ != nullptr)
        {
            StateVarsCompartment *nextCompartment = this->_nextCompartment_;
            this->_nextCompartment_ = nullptr;
            if (nextCompartment->_forwardEvent != nullptr && 
                nextCompartment->_forwardEvent->_message == ">")
            {
                this->_mux_(new FrameEvent( "<", this->_compartment_->exitArgs));
                this->_compartment_ = nextCompartment;
                this->_mux_(nextCompartment->_forwardEvent);
            }
            else
            {
                this->_doTransition_(nextCompartment);
                if (nextCompartment->_forwardEvent != nullptr)
                {
                    this->_mux_(nextCompartment->_forwardEvent);
                }
            }
            nextCompartment->_forwardEvent = nullptr;
        }
    }
    
        
        //===================== Interface Block ===================//
    public:
        void X() {
            FrameEvent e("X", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void Y() {
            FrameEvent e("Y", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void Z() {
            FrameEvent e("Z", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
    
    //===================== Machine Block ===================//

private:
    
    void _sInit_(FrameEvent *e)
    {
        if (e->_message == ">") {
            StateVarsCompartment *compartment =  new StateVarsCompartment(static_cast<int>(StateVarsState::A));
            compartment->stateVars["x"] = 0;
            
            this->_transition_(compartment);
            return;
        }
    }
    
    void _sA_(FrameEvent *e)
    {
        if (e->_message == "X") {
            this->_compartment_->stateVars["x"] = (any_cast<int>(this->_compartment_->stateVars["x"])) + 1;
            return;
        }
        else if (e->_message == "Y") {
            StateVarsCompartment *compartment =  new StateVarsCompartment(static_cast<int>(StateVarsState::B));
            compartment->stateVars["y"] = 10;
            compartment->stateVars["z"] = 100;
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Z") {
            StateVarsCompartment *compartment =  new StateVarsCompartment(static_cast<int>(StateVarsState::B));
            compartment->stateVars["y"] = 10;
            compartment->stateVars["z"] = 100;
            
            this->_transition_(compartment);
            return;
        }
    }
    
    void _sB_(FrameEvent *e)
    {
        if (e->_message == "X") {
            StateVarsCompartment *compartment =  new StateVarsCompartment(static_cast<int>(StateVarsState::A));
            compartment->stateVars["x"] = 0;
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Y") {
            this->_compartment_->stateVars["y"] = (any_cast<int>(this->_compartment_->stateVars["y"])) + 1;
            return;
        }
        else if (e->_message == "Z") {
            this->_compartment_->stateVars["z"] = (any_cast<int>(this->_compartment_->stateVars["z"])) + 1;
            return;
        }
    }

//===================== Actions Block ===================//
    
    
    // Unimplemented Actions
    public:
    
    //===================== Domain Block ===================//
    
    
    
    
    //=============== Machinery and Mechanisms ==============//

private:
    int _state_;
    
    void _transition_(StateVarsCompartment *compartment)
    {
        _nextCompartment_ = compartment;
    }
    
    void _doTransition_(StateVarsCompartment *nextCompartment)
    {
        this->_mux_(new FrameEvent("<", _compartment_->exitArgs));
        this->_compartment_ = nextCompartment;
        this->_mux_(new FrameEvent(">", this->_compartment_->enterArgs));
    }
    
    public:
    string state_info(){
        return std::to_string(_compartment_->state);
        }
        
};

/********************

class StateVarsController : public StateVars
{
public:
	StateVarsController() : StateVars() {}
};

********************/

